Export symbols in Module even with MINIMAL_RUNTIME

DavidPeicho opened the pull request:
This is an attempt to fix: #17852

## Why

Right now when creating a WASM module that can be managed by the JS (whether it's a library, or just a game with data coming from the JS), symbols aren't accessible outside the module when using `MODULARIZE`.

We could hide this change behind another flag like `STANDALONE` vs `LIBRARY`, but that's yet another flag to support.

DavidPeicho commented:
It fails because obviously the size aren't matching anymore. How do you want to proceed @kripken and @juj?

sbc100 commented on emscripten.py:
```diff
@@ -737,7 +737,7 @@ def create_receiving(exports):
         mangled = asmjs_mangle(s)
         dynCallAssignment = ('dynCalls["' + s.replace('dynCall_', '') + '"] = ') if generate_dyncall_assignment and mangled.startswith('dynCall_') else ''
-        receiving += [dynCallAssignment + mangled + ' = asm["' + s + '"];']
+        receiving += [dynCallAssignment + mangled + ' = Module["{}"] = asm["{}"];'.format(mangled, s)]
```
I think this should only be in he `MODULARIZE` case.

DavidPeicho replied:
Done 👌

sbc100 commented on src/preamble_minimal.js:
```diff
@@ -65,17 +65,17 @@ function updateGlobalBufferAndViews(b) {
-  HEAP64 = new BigInt64Array(b);
-  HEAPU64 = new BigUint64Array(b);
+  Module['HEAP64'] = HEAP64 = new BigInt64Array(b);
+  Module['HEAPU64'] = HEAPU64 = new BigUint64Array(b);
```
Ditto, the assignments to Module should only be needed in the MODULARIZE case.

DavidPeicho replied:
Looking at `preamble.js`, it happens no matter if it's a module or not. So I guess we should do the same?

sbc100 replied:
I believe one of main differences with MINIMAL_RUNTIME is that it does away with the Module object by default.  To save space.   With MODULARIZE however, we are force to revert to exporting everything on that object.

sbc100 reviewed (commented):
Can we add test for this?

sbc100 commented:
Can you mention MODULARIZE in the title since this probably only effects that mode, right?

sbc100 reviewed (commented):
Great!  Would be good to have a test for this but otherwise lgtm

DavidPeicho commented:
Wondering what kind of tests we can do, just something that test that the function pointers are exported and available in JS?
What should we do for the size test that's failing?

kripken commented:
About code size, should we be exporting these symbols by default? Perhaps we should only export the ones in `EXPORTED_FUNCTIONS`, that is, the user must opt in to it. I believe that is the goal of @juj's design for minimal runtime, to not add extra code size by default ever, and make all increases opt-in.

DavidPeicho commented:
I feel like it should be exported, as per the doc:
> EMSCRIPTEN_KEEPALIVE also exports the function, as if it were on EXPORTED_FUNCTIONS

Otherwise using `EMSCRIPTEN_KEEPALIVE` just means preventing inlining 🫤

juj commented:
Sorry for the delay in review, I have been away for two weeks, slowly unpiling from review queue now.

The original idea with MINIMAL_RUNTIME was to stop exporting any symbols by default, since these exports produce a significant source of bloat. For MINIMAL_RUNTIME we are mindful about every single byte of generated code (this includes MINIMAL_RUNTIME + MODULARIZE combination), since that works best for ensuring Emscripten is good for all sorts of target deployments.

For module exports in MINIMAL_RUNTIME, the thinking was to migrate to a model where the developer could from command line define an exports list. I recall some of the ideas were originally thrown around in the original MINIMAL_RUNTIME PR ticket. To recap, something like

```
em++ -sMINIMAL_RUNTIME mycode.cpp -o out.html --export exports.list
```

where the file could define with glob/regex syntax a set of symbols that are desired to be exported out. This way developers would be able to explicitly control which symbols would constitute their public API, and avoid piling all kinds of symbols on the `Module` object (which is unfortunately quite overloaded in meaning already as it is).

The `--export` command could take in either a comma-separated list, or a filename, so developers who would want to simply "just export everything" would be able to pass `--export *`, or a quick `--export func1,func2,func3`.

However people have mostly used MINIMAL_RUNTIME specifically for self-contained micro-sized apps, so the whole `--export` feature did not become urgent to implement, hence the feature now missing altogether. Now might be a good timing to revisit that.

There were further ideas about the `exports.list` being able to define a shape of an exports API, something like

```js
Module.start = main;
Module.img.new = img_new;
Module.img.load = img_load;
Module.img.save = img_save;
```
that would allow one to rename functions that are exported out to JS for clearer meaning, and also create an export style object (and also integrate with ES6 exports statement if targeting that), and be able to structure the exported function in the definition. Though not sure if there would be many users for this kind of structuring, that was just spitballing ideas in case that might be useful.

I suppose current `--post-js` feature could already today be used to define exports for MINIMAL_RUNTIME?

sbc100 commented:
> This way developers would be able to explicitly control which symbols would constitute their public API, and avoid piling all kinds of symbols on the `Module` object (which is unfortunately quite overloaded in meaning already as it is).

I like the idea of discussing a possible new `--export=filename` flag and the possibility of also mapping internal to external names, I don't quite understand why we can't make the current export methods just work for MINIMAL_RUNTIME.   I think alon proposed something like this in https://github.com/emscripten-core/emscripten/issues/8380.

We have two methods of specifying exports today and they both only export the precise list of the symbols the user asks to export:

1. -sEXPORTED_FUNCTIONS
2. EMSCRIPTEN_KEEPALIVE

In programs that don't specify either of these they would be no symbols exported at all.

It sounds like you are worried about accidentally exporting everything.. which is not something that we support at all today as far as I know.

(BTW `-sEXPORTED_FUNCTIONS` also supports the `@filename` syntax if you want to put your symbol list in a file, although there is no whildcard support).

@juj would be you be ok with this change/direction if it doesn't regress any code size tests by a single byte?

DavidPeicho commented:
To me the `export` list sounds good, but it's another issue. Right now `EMSCRIPTEN_KEEPALIVE` doesn't export the symbol at all in MINIMAL_RUNTIME. I think people building small standalone should not use it in order to get no symbol exported.

juj commented:
> We have two methods of specifying exports today and they both only export the precise list of the symbols the user asks to export

Sorry, I find now I have missed a change here. Indeed you are correct. Making `-sEXPORTED_FUNCTIONS` and `EMSCRIPTEN_KEEPALIVE` work for MINIMAL_RUNTIME does make sense if we can do it in a way that does not regress code size for people who don't need any exports.

> It sounds like you are worried about accidentally exporting everything.. which is not something that we support at all today as far as I know.

That's correct. When MINIMAL_RUNTIME landed, we had an issue with a lot of redundant stuff getting exported, but now I see that is no longer the case.

Though I do also worry not just about exporting everything, but exporting anything that is not needed.

Note that today, there are two overloaded meanings for the word exporting. There is the exporting from inside Wasm code out to JS code. And then there is the exporting from JS code out to "external world", i.e. `Module` (or to ES6 exports).

If we had a way to resolve issue #15211, then distinguishing between these two meanings (for end users) could be collapsed into one, and exporting (for end users) could be always used in the meaning "export to external world". But because of issue #15211 users may need to export a symbol just so that their own JS libraries can access a wasm symbol, since otherwise they don't have a way of saying "this JS function depends on that wasm function".

juj commented:
> @juj would be you be ok with this change/direction if it doesn't regress any code size tests by a single byte?

Yeah, that would definitely be great to land!

DavidPeicho commented:
So how do we approach that? In the end I just wish I don't need to do anything else than adding something above my exported function.

sbc100 commented:
> So how do we approach that? In the end I just wish I don't need to do anything else than adding something above my exported function.

We have code size tests that you can run using:

`$ ./tests/runner other.*code_size* other.*metadce*`

If you can all those tests without changing the output sizes that you know you change is probably code-size safe.

You can use `--rebase` to update the expected outout.  My normal workflow is this:

```
1. Checkout origin/main
2. Run  `./tests/runner other.*code_size* other.*metadce* --rebase` and submit the results (to make sure the results are fully accurate on `main`
3. Rebase my branch of top of the updartes in (2)
4. Re-run `./tests/runner other.*code_size* other.*metadce* --rebase` and make sure no files were changed
```

kripken commented:
I'm not sure I follow the last part of the conversation. It seems like if we made `EMSCRIPTEN_KEEPALIVE` work with `MINIMAL_RUNTIME` then that would give us what we want? Is the tricky part implementing that, or avoiding code size regressions while doing that? (or something else?)

sbc100 commented:
> I'm not sure I follow the last part of the conversation. It seems like if we made `EMSCRIPTEN_KEEPALIVE` work with `MINIMAL_RUNTIME` then that would give us what we want? Is the tricky part implementing that, or avoiding code size regressions while doing that? (or something else?)

I think the reason it won't work easily today is that we don't have `EMSCRIPTEN_KEEPALIVE_FOR_EXTERNAL_USE` and `EMSCRIPTEN_KEEPALIVE_FOR_INTERNAL_USE`. 

For the non-minimal runtime we assume `EMSCRIPTEN_KEEPALIVE` is for external use so such symbols get exported on the Module.. which is an unnecessary overhead if it was only intended for internal use.

DavidPeicho commented:
Should we add `EMSCRIPTEN_KEEPALIVE_FOR_EXTERNAL_USE`? It sounds like another hard thing to maintain.

sbc100 commented:
> Should we add `EMSCRIPTEN_KEEPALIVE_FOR_EXTERNAL_USE`? It sounds like another hard thing to maintain.

Or we could just call it `EMSCRIPTEN_EXPORT`..  the problem with doing that is that pretyt much all our users are already using `EMSCRIPTEN_KEEPALIVE` to mean "export for external use" so changing the meaning of the existing macro would be hard.

DavidPeicho commented:
I am up to use the KEEPALIVE for export

DavidPeicho commented:
How big of a task would it be to add this `EMSCRIPTEN_EXPORT`? Or at least a flag that would treat `EMSCRIPTEN_KEEPALIVE` as an export token as well. The latter might be easier and would fit in the few lines of code I wrote here.

sbc100 commented:
I think a flag to make `EMSCRIPTEN_KEEPALIVE` behave the same in minimal runtime mode would probably make sense, since `EMSCRIPTEN_KEEPALIVE` has always meant "export to the outside world" in that past, it seems reasonable to able to keep that behaviour.

sbc100 commented:
Adding a separate `EMSCRIPTEN_EXPORT` requires a patch to llvm I believe, so it more work to make that work.

sbc100 commented:
> I was working on that right now, and one thing I am unsure, why can't `EXPORT_ALL` do that? To me, having `EXPORT_ALL` + `MODULARIZE` + `MINIMAL_RUNTIME` should perform the task of:
> 
> ```js
> Module[name] = asm[name]
> ```
> 
> Maybe I am confusing what this option does? To me it looks like it

I agree that could make sense to implement.   I imagine that `EXPORT_ALL` has no meaning at all today with `MINIMAL_RUNTIME`.

One problem with `EXPORT_ALL` is that its not selective and will just export everything, rather then just what the user wants.   But I don't see why we shouldn't make it work.

DavidPeicho commented:
Okay I will try with `EXPORT_ALL` instead of adding a new option that would make it really hard to understand what options to choose depending on `MODULARIZE` / `MINIMAL_RUNTIME`.

DavidPeicho commented:
@sbc100 I gave it another go, using `EXPORT_ALL`. I added a test that passes.

I also modified `modules.js` that would cause issue: Basically the unexported getter check would be setup before the exports are setup on `Module` (see [here](https://github.com/emscripten-core/emscripten/pull/17911/files#diff-7cfe3e758b9e85c72272acc2ef6899b684e5e943f17ba18b52a9af8b587e95f8)). A better fix would be to actually only export the things that should really be exported, but I wasn't sure how to do that. In general anyway, even if it's maybe another discussion, all those safety checks should maybe not be performed in `MINIMAL_RUNTIME` due to the amount of code it adds.

sbc100 commented:
> In general anyway, even if it's maybe another discussion, all those safety checks should maybe not be performed in `MINIMAL_RUNTIME` due to the amount of code it adds.

The safety changes are presumably only added in ASSERTIONS mode, in which case the size of the code is not so important.

DavidPeicho commented:
We either still need to keep my change, or make a better export wrapper, otherwise someone with `ASSERTIONS` mode and `MODULARIZE` and `MINIMAL_RUNTIME` will have an exception.

kripken commented on src/preamble_minimal.js:
```diff
@@ -77,6 +77,24 @@ function updateMemoryViews() {
   HEAPU64 = new BigUint64Array(b);
 #endif
+
+#if MODULARIZE
+#if EXPORT_ALL
```
```suggestion
#if MODULARIZE && EXPORT_ALL
```

kripken commented on src/preamble_minimal.js:
```diff
@@ -77,6 +77,24 @@ function updateMemoryViews() {
+  Module['HEAP64'] = HEAP64;
+  Module['HEAPU64'] = HEAPU64;
+#endif
+#endif
+#endif
```
```suggestion
#endif
```

DavidPeicho replied:
Done!

kripken commented on test/test_other.py:
```diff
@@ -1306,6 +1306,42 @@ def test_export_all(self):
 
+  def test_export_all_modularize(self):
+    """
+    This test ensures that modularize and export all work simulatenously.
```
```suggestion
    This test ensures that modularize and export all work simultaneously.
```

kripken commented on test/test_other.py:
```diff
@@ -1306,6 +1306,42 @@ def test_export_all(self):
     self.emcc('lib.c', ['-Oz', '-sEXPORT_ALL', '-sLINKABLE', '--pre-js', 'main.js'], output_filename='a.out.js')
     self.assertContained('libf1\nlibf2\n', self.run_js('a.out.js'))
 
+  def test_export_all_modularize(self):
```
```suggestion
  def test_minimal_runtime_export_all_modularize(self):
```

DavidPeicho replied:
Done!

kripken commented on test/test_other.py:
```diff
@@ -1306,6 +1306,42 @@ def test_export_all(self):
+
+    self.emcc('main.c', ['-sMODULARIZE=1', '-sMINIMAL_RUNTIME=2', '-sEXPORT_ALL', '-sEXPORT_ES6'], output_filename='test.mjs')
+
+    # We must expose __dirname and require globally because emscripten
+    # uses those under the hood.
```
Why is this an issue in this test but not in general?

DavidPeicho replied:
Basically this test runs in nodejs and import a module, however using `MODULARIZE` and `MINIMAL_RUNTIME`, the UMD export is disabled. I exported using `ES6` and it looks like this doesn't run directly in nodejs due to the internal code using require (not available in an ES6 module in node by default).

I did a quick search and the other tests are either running in the web, or are just checking the content of the file written.

If there is a better way to run the test, maybe in the browser, and get the output, I don't mind doing so.

kripken replied:
> I did a quick search and the other tests are either running in the web, or are just checking the content of the file written.

"the other tests" - which subset of tests do you mean? (99% of our tests run in node, so you must need some specific feature here?)

DavidPeicho replied:
The tests already written don't do `MODULARIZE` + `MINIMAL_RUNTIME` + `EXPORT_ES6` in node AFAIK, if I am wrong can you please point me to one?
The tests that mix `MODULARIZE` and `MINIMAL_RUNTIME` do not import nor require the exported module (example of test like that: `browser.test_minimal_runtime_modularize`)

sbc100 replied:
Yes, that looks like current situation.  We don't test `MODULARIZE` + `MINIMAL_RUNTIME` except in browser tests.  I was looking at this just yesterday.  Not sure how hard it would be to do some testing under node too..  not completely trivial but probably worth looking into

kripken replied:
Good ideas. Yes, it would be nice to test modularize + minimal runtime outside of the browser, and maybe we can extract some helper code for that from this test later.

kripken reviewed (commented):
Direction here lgtm!

DavidPeicho commented:
There are still two little things I am a bit skeptical about and you might have better ideas:
* The `HEAP` is going to be exported only using `EXPORT_ALL`. So, someone manually selecting what to export will not have it. Not sure what would be the best way to fix that.
* I would have like actually that the assertions work with the module exports, because it could be nice for debugging purposes. That looks a little annoying to do though

juj commented on src/modules.js:
```diff
@@ -447,7 +447,7 @@ function exportRuntime() {
     }
 
-    if (unexported.length) {
+    if (!MINIMAL_RUNTIME && unexported.length) {
```
This block runs inside `ASSERTIONS` builds, i.e. it is intended to help at debug time to catch errors. So it would be ok to enable that for `MINIMAL_RUNTIME` as well to help debug issues there. Or was there some kind of issue emitting this code into `MINIMAL_RUNTIME` builds?

DavidPeicho replied:
The issue is that when we start exporting the symbols, the export is done after this getter is installed, so:

```js
// Install getter here.
...
// Export on `Module`
Module[_my_exported_function] = ...; // Throws an error
```

In general as we mentioned throughout this PR, there are issues considering what's exported for usage inside the module, and what's exported to the outside world. So we could either find a way to fixup this `unexported` list, or install the assertion later in the process when using `MINIMAL_RUNTIME`.

DavidPeicho replied:
Good point, added a comment 👌

juj commented on src/preamble_minimal.js:
```diff
@@ -77,6 +77,22 @@ function updateMemoryViews() {
+  Module['HEAP64'] = HEAP64;
+  Module['HEAPU64'] = HEAPU64;
+#endif
+#endif
```
With a little bit of magic with the `{{{ }}}` syntax, the above can be unified into

```diff
diff --git a/src/preamble_minimal.js b/src/preamble_minimal.js
index 2f32865d9..56e96df31 100644
--- a/src/preamble_minimal.js
+++ b/src/preamble_minimal.js
@@ -62,20 +62,28 @@ function updateMemoryViews() {
 #if ASSERTIONS && SHARED_MEMORY
   assert(b instanceof SharedArrayBuffer, 'requested a shared WebAssembly.Memory but the returned buffer is not a SharedArrayBuffer, indicating that while the browser has SharedArrayBuffer it does not have WebAssembly threads support - you may need to set a flag');
 #endif
+
+{{{ (function() {
+  global.assignToModuleIfExportAll = function(x) {
+    return MODULARIZE && EXPORT_ALL ? `Module['${x}'] =` : '';
+  };
+  return null;
+})(); }}}
+
 #if SUPPORT_BIG_ENDIAN
-  HEAP_DATA_VIEW = new DataView(b);
-#endif
-  HEAP8 = new Int8Array(b);
-  HEAP16 = new Int16Array(b);
-  HEAP32 = new Int32Array(b);
-  HEAPU8 = new Uint8Array(b);
-  HEAPU16 = new Uint16Array(b);
-  HEAPU32 = new Uint32Array(b);
-  HEAPF32 = new Float32Array(b);
-  HEAPF64 = new Float64Array(b);
+  {{{ assignToModuleIfExportAll('HEAP_DATA_VIEW') }}} HEAP_DATA_VIEW = new DataView(b);
+#endif
+  {{{ assignToModuleIfExportAll('HEAP8') }}} HEAP8 = new Int8Array(b);
+  {{{ assignToModuleIfExportAll('HEAP16') }}} HEAP16 = new Int16Array(b);
+  {{{ assignToModuleIfExportAll('HEAP32') }}} HEAP32 = new Int32Array(b);
+  {{{ assignToModuleIfExportAll('HEAPU8') }}} HEAPU8 = new Uint8Array(b);
+  {{{ assignToModuleIfExportAll('HEAPU16') }}} HEAPU16 = new Uint16Array(b);
+  {{{ assignToModuleIfExportAll('HEAPU32') }}} HEAPU32 = new Uint32Array(b);
+  {{{ assignToModuleIfExportAll('HEAPF32') }}} HEAPF32 = new Float32Array(b);
+  {{{ assignToModuleIfExportAll('HEAPF64') }}} HEAPF64 = new Float64Array(b);
 #if WASM_BIGINT
-  HEAP64 = new BigInt64Array(b);
-  HEAPU64 = new BigUint64Array(b);
+  {{{ assignToModuleIfExportAll('HEAP64') }}} HEAP64 = new BigInt64Array(b);
+  {{{ assignToModuleIfExportAll('HEAPU64') }}} HEAPU64 = new BigUint64Array(b);
 #endif
 }
```
to improve code size.

juj replied:
> The HEAP is going to be exported only using EXPORT_ALL. So, someone manually selecting what to export will not have it. Not sure what would be the best way to fix that.

So far in MINIMAL_RUNTIME there does not exist any kind of mechanism to export out to `Module`, so this `EXPORT_ALL` flag will be the first mechanism to enable that for `MINIMAL_RUNTIME`. But if we do build such a method in the future, then the above type of `assignToModuleIfExportAll()` function would be able to look into an exports array per each symbol to see which assignments to outside world it would want to do.

DavidPeicho replied:
I like this idea, done :)

juj commented on emscripten.py:
```diff
@@ -753,7 +753,11 @@ def create_receiving(exports):
+        if settings.MODULARIZE and settings.EXPORT_ALL:
+          receiving += [dynCallAssignment + mangled + ' = Module["{}"] = asm["{}"];'.format(mangled, s)]
+        else:
+          receiving += [dynCallAssignment + mangled + ' = asm["' + s + '"];']
```
Maybe this could be
```
exportAssignment = f'Module["{mangled}"] =` if settings.MODULARIZE and settings.EXPORT_ALL else ''
receiving += [exportAssignment + dynCallAssignment + mangled + ' = asm["' + s + '"];']
```
to follow the same kind of pattern that `dynCallAssignment` uses?

DavidPeicho replied:
Yup good idea, I made a better version 👌

DavidPeicho commented:
Thanks a lot for the review @juj. Anything missing here? :)

juj commented on emscripten.py:
```diff
@@ -753,7 +753,8 @@ def create_receiving(exports):
         dynCallAssignment = ('dynCalls["' + s.replace('dynCall_', '') + '"] = ') if generate_dyncall_assignment and mangled.startswith('dynCall_') else ''
-        receiving += [dynCallAssignment + mangled + ' = asm["' + s + '"];']
+        export_assignment = 'Module["{}"] = asm["{}"];'.format(mangled, s) if settings.MODULARIZE and settings.EXPORT_ALL else 'asm["{}"]'.format(s)
+        receiving += [dynCallAssignment + mangled + ' = {};'.format(export_assignment)]
```
Hmm, is there a specific reason to have the `asm['x']` part be formatted on both sides of the branch?

Now that we are fully Python 3, I would recommend using f-strings over `.format()` and `%`, e.g.

```
export_assignment = f'Module["{mangled}"] =' if settings.MODULARIZE and settings.EXPORT_ALL else ''
receiving += [f'{export_assignment} {dynCallAssignment} {mangled} = asm["{s}"]']
```
might read the cleanest without duplication?

DavidPeicho replied:
Done, use fstrings

juj commented:
LGTM, leaving for @kripken for final review.

DavidPeicho commented:
Not sure why the tests are red they run green locally in chrome. I might need to try the exact same version.

kripken commented:
As a last request, please add something to `Changelog.md`, as this is a useful improvement that seems worth mentioning. Thanks!

sbc100 commented on src/preamble_minimal.js:
```diff
@@ -56,26 +56,32 @@ var HEAP8, HEAP16, HEAP32, HEAPU8, HEAPU16, HEAPU32, HEAPF32, HEAPF64,
+    return MODULARIZE && EXPORT_ALL ? `Module['${x}'] =` : '';
+  };
+  return null;
+})(); }}}
```
See the pattern in `src/library_html5_webgpu.js ` for how to do this without the IIFE

sbc100 commented on src/modules.js:
```diff
@@ -447,7 +447,10 @@ function exportRuntime() {
+    // When using MINIMAL_RUNTIME, MODULARIZE, and EXPORT_ALL,
+    // the symbols will be assigned to the `Module` object, but only **after**
+    // this code.
+    if ((!MINIMAL_RUNTIME || !EXPORT_ALL) && unexported.length) {
```
I think we still want to support unexportedRuntimeSymbols, so we should find a way to restructure or reorder such that this keeps working.

DavidPeicho replied:
Do we need to in case? Because it's specifically disabled for `EXPORT_ALL`.

sbc100 replied:
Oh I see, perhaps the whole purpose of unexportedRuntimeSymbols is mute in `EXPORT_ALL` mode.

Perhaps we should change line 434 above to `if (ASSERTIONS && !EXPORT_ALL) {` above?   That might be good to land as a separate change with a separate test since it doesn't just effect minimal runtime.  As part of that change we could also then wrap the declaration of `unexportedRuntimeSymbols` itself if `if (!EXPORT_ALL)`

sbc100 replied:
(I can take care of that you like).

DavidPeicho replied:
Maybe the first warning is actually useful?

sbc100 replied:
Yes, I'm actually not sure why that that first warning was inside this block at all.  Fixed that in #18567

DavidPeicho replied:
Okay great, I will rebase this PR. Do you want to make the change for export all or should I do it?

sbc100 commented on emscripten.py:
```diff
@@ -756,7 +756,8 @@ def create_receiving(exports):
         dynCallAssignment = ('dynCalls["' + s.replace('dynCall_', '') + '"] = ') if generate_dyncall_assignment and mangled.startswith('dynCall_') else ''
-        receiving += [dynCallAssignment + mangled + ' = asm["' + s + '"];']
+        export_assignment = f'Module["{mangled}"] = ' if settings.MODULARIZE and settings.EXPORT_ALL else ''
+        receiving += [f'{export_assignment}{dynCallAssignment}{mangled} = asm["{s}"]']
```
It seems like maybe we a duplicating code from `make_export_wrappers` here.. can we just call it instead?

I'm not also no a big fan of the use of `if else` like this on a single line.  How about:

```
if settings.MODULARIZE and settings.EXPORT_ALL:
  export_assignment = f'Module["{mangled}"] = '
else:
  export_assignment = ''
``

DavidPeicho replied:
For the `if else` on the same line, it was requested in an earlier comment to keep it similar to the code there was before (the `dynCallAssignment`.

I don't know if we want to call `make_export_wrappers` because in the case of `MINIMAL_RUNTIME`, the `var` definition is done ahead of this function. I technically can move this code in the `make_export_wrappers` if you want, and call it with `exports_that_are_not_initializers`.

juj replied:
fwiw I think the single-line code is more readable.

sbc100 replied:
I do feel a bit more strongly about these.  Personally, I find maintaining code that uses this style a lot harder.  There seem to be general agreement that its bad style here: https://stackoverflow.com/questions/35070436/python-ternary-style-is-this-good-or-bad-style.   

The google style guide does say they are OK in simple cases:  https://google.github.io/styleguide/pyguide.html#2114-decision.   But I would say when the length of the line is over 80 chars, and contains multiple embedded f-strings, we a no longer that catagory.

DavidPeicho replied:
I pushed without any `else` so I think it meets halfway 😄

sbc100 commented on src/preamble_minimal.js:
```diff
@@ -56,26 +56,32 @@ var HEAP8, HEAP16, HEAP32, HEAPU8, HEAPU16, HEAPU32, HEAPF32, HEAPF64,
-  HEAPU32 = new Uint32Array(b);
-  HEAPF32 = new Float32Array(b);
-  HEAPF64 = new Float64Array(b);
+  {{{ assignToModuleIfExportAll('HEAP_DATA_VIEW') }}} HEAP_DATA_VIEW = new DataView(b);
```
How about just calling this `maybExport`?

juj commented:
Sorry for mixed reviews, though I'd say `maybeExport` is at least an order of magnitude worse than `assignToModuleIfExportAll()`. I don't think we should be using the word `maybe` in almost any identifier anywhere ever. It feels similar to naming a function `doStuff()`, kind of explicitly highlighting the reader "I couldn't convey what this does, figure it out yourself".

Though I do see it is shorter name at least, and don't want to create any churn with nameshedding, so lgtm with this :)

DavidPeicho commented:
I personnally don't mind too much, I just looked in the codebase before and it's used pretty often for a lot of stuff. So I guess it's fair to keep it like that and maybe make a rename if needed that's more global.

DavidPeicho commented:
There are new comments but not sure if we want to tackle those here. In general, we are modularizing + using minimal runtime in our codebase, so any further missing feature or issue is likely to be found pretty soon on my side as well.

sbc100 commented:
I've been working a couple of related changes (not specific to MINIMAL_RUNTIME).  One is to reduce the size of the code that `create_export_wrappers` creates: #18564.  A another is to stop using `EXPORT_ALL` with dynamic linking.   I was kind of hoping that this would allow the complete removal of the `EXPORT_ALL` option, but perhaps we will need to keep it around for this purpose.    A big downside of EXPORT_ALL is that if mean we are severely limited in the DCE we can perform.   I assume a more precise method of choosing exports would be preferable (in the long run) for everyone?  Is that true for you too @DavidPeicho?

All the is to say I don't mind this change landing (assuming we can resolve open issues), but I would hope in the long run we wouldn't need to resort of `EXPORT_ALL`.

DavidPeicho commented:
I don't mind not having `EXPORT_ALL` but I don't want to list of my C functions that are written as `KEEPALIVE`, so it's a bit of a circle back to a `EMSCRIPTEN_EXPORT` macro (that could indeed land later).

sbc100 commented:
> I don't mind not having `EXPORT_ALL` but I don't want to list of my C functions that are written as `KEEPALIVE`, so it's a bit of a circle back to a `EMSCRIPTEN_EXPORT` macro (that could indeed land later).

Agreed.   Anther option would be something like `EXPORT_KEPT_ALIVE` (probably with a better name than that!), which would be less aggressive/conservative than `EXPORT_ALL`

DavidPeicho commented:
I think I started on this idea, but was a little more complex. I think it's indeed cleaner, as long as we don't forget to expose the `HEAP` variable as well as I did (pretty useful). How hard would it be to implement that? I was kind of hopping we could get this merge before the next release so I can start the modularization on my side 😄

sbc100 commented:
> Sorry for mixed reviews, though I'd say `maybeExport` is at least an order of magnitude worse than `assignToModuleIfExportAll()`. I don't think we should be using the word `maybe` in almost any identifier anywhere ever. It feels similar to naming a function `doStuff()`, kind of explicitly highlighting the reader "I couldn't convey what this does, figure it out yourself".
> 
> Though I do see it is shorter name at least, and don't want to create any churn with nameshedding, so lgtm with this :)

Sure, I can see that argument for a longer more descriptive name.  I tend to prefer short names for locally-defined-and-used things.   I don't feel strongly if this case, feel free to go back to the longer name if you prefer it.

DavidPeicho commented:
I made the `if else` change, but kept the `maybeExport` because I think it matches the rest of the codebase. Not to vocal about it either, but looks like it's fine to keep it as-is.

DavidPeicho commented:
Rebased, and make the fix with `ASSERTION && !EXPORT_ALL`. Do you guys think we can merge that?

sbc100 commented on test/test_other.py:
```diff
@@ -1307,6 +1307,41 @@ def test_export_all(self):
+      EMSCRIPTEN_KEEPALIVE void libf1() { printf("libf1\n"); }
+      EMSCRIPTEN_KEEPALIVE void libf2() { printf("libf2\n"); }
+    '''
+    create_file('main.c', main)
```
You can inline the string into `create_file`.  e.g. `create_file('main.c', '''...`

sbc100 commented on test/test_other.py:
```diff
@@ -1307,6 +1307,41 @@ def test_export_all(self):
 
+  def test_minimal_runtime_export_all_modularize(self):
+    """
+    This test ensures that modularize and export all work simultaneously.
```
This line of docstring goes on same line as opening """

sbc100 replied:
Capitalize `MODULARIZE` and `EXPORT_ALL` here.

sbc100 commented on src/preamble_minimal.js:
```diff
@@ -56,26 +65,25 @@ var HEAP8, HEAP16, HEAP32, HEAPU8, HEAPU16, HEAPU32, HEAPF32, HEAPF64,
-  HEAP64 = new BigInt64Array(b);
-  HEAPU64 = new BigUint64Array(b);
+  {{{ maybeExport('HEAP64') }}} HEAP64 = new BigInt64Array(b);
+  {{{ maybeExport('HEAPU64') }}} HEAPU64 = new BigUint64Array(b);
```
Avoid adding an extra space between the template the declaration, and instead add the space after the `=` withing the template on line 11

sbc100 commented on ChangeLog.md:
```diff
@@ -31,6 +31,8 @@ See docs/process.md for more on how version tagging works.
   like JS library files and the core runtime JS files.  This means they can
   now contain #if/#else/#endif blocks and {{{ }}} macro blocks. (#18525)
+- `-sEXPORT_ALL` can now be used to export symbols on the `Module` object
+  when used with `-sMINIMA_RUNTIME` and `-sMODULARIZE` together.
```
Add `(#17911)` here.

sbc100 reviewed (approved):
LGTM with a couple of nits

DavidPeicho commented:
@sbc100 Anyway to re-run the windows job? The install fails from time to time not sure why

DavidPeicho commented:
Would be nice yes I don't have the rights to re-start it :)

dschuff commented:
This test seems to be failing on Windows on Chromium CI (https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8790978322466379425/+/u/Emscripten_testsuite__other_/stdout) which is weird because I don't see any way this should be OS-dependent, and the error in the log isn't clear to me. But it is failing reliably.

sbc100 commented:
I think the problem is the module loading code used in the test is probably not portable.  Maybe this code:

```
globalThis.__dirname = dirname(import.meta.url).substring(7);
globalThis.require = createRequire(import.meta.url);
```

dschuff commented:
Oh, yeah... even if it were portable it's not at all obvious what that's supposed to be cutting off?

sbc100 commented:
I think the solution is to run add this test to mac/windows github CI .. and maybe disable it under windows until it can be fixed.

DavidPeicho commented:
Sorry for that, not sure how I ended up with the `substring` but it's definitely not correct. I will make a new PR to fix this

DavidPeicho commented:
So the error is the `__dirname` used as a base: `'/C:/b/s/w/ir/x/t/emtest_6n138uii/emscripten_test_other_z3ln08yu/test.wasm'` There is an extra `/`.

MR opened here: https://github.com/emscripten-core/emscripten/pull/18601
